### PR TITLE
adding okhttp interface

### PR DESCRIPTION
**Pull request by EvangelosG**
@sojanpr

### CODE REVIEW

File: Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                    retryNumber < prefHelper.getRetryCount()) {
+                try {
+                    Thread.sleep(prefHelper.getRetryInterval());
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+                retryNumber++;
+                return doRestfulGet(url, retryNumber);
+            } else {
+                return new BranchResponse(getResponseString(response.isSuccessful() ? response.body().byteStream() : null), responseCode);
```
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java]: Need null check before accessing `response.body()`

File: Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                    && retryNumber < prefHelper.getRetryCount()) {
+                try {
+                    Thread.sleep(prefHelper.getRetryInterval());
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+                retryNumber++;
+                return doRestfulPost(url, payload, retryNumber);
+            } else {
+                return new BranchResponse(getResponseString(response.isSuccessful() ? response.body().byteStream() : null), responseCode);
```
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java]: Same  null check

**Comment by sojanpr** [comment]
@EvangelosG  How are we handling the `retry` here?  It looks like  we have enable `retryOnConnectionFailure` wit OKHttp and the we again `Retry` ourself.  Cann't we set the retry policy and interval with OKhttp?

File: Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        } catch (SocketTimeoutException ex) {
```
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java]: Looks like OKhttp throes only IOException or IllegalState Exception

**Comment by sojanpr** [comment]
OKhttp doc suggest ```To avoid leaking resources callers should close the {@link Response} which in turn will close the underlying {@link ResponseBody}.```. Not seeing we are closing it any where. Is it internally handled somehow?

**Comment by EvangelosG** [comment]
thanks! added finally blocks to close response. this call was included in 0 of their ready-to-go recipes..

File: Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java (hunk 3 above)
Comment (EvangelosG) [inline Branch-SDK/src/io/branch/referral/network/BranchRemoteInterfaceOkHttp.java]: great, removed those blocks

**Comment by EvangelosG** [comment]
also removed okhttp's handling of retries and kept our code to honor retryCount

**Comment by sojanpr** [comment]
Closing this out since SDK provide a generic  interface for custom network handling here.
https://github.com/BranchMetrics/android-branch-deep-linking/blob/master/Branch-SDK/src/io/branch/referral/network/BranchRemoteInterface.java

**Pull request closed without merging**